javers-core/src/main/java/org/javers/snapshots-compiler/JaversSnapshotsCompiler.java

```diff
@@ -0,0 +1,216 @@
+            Object propertyValue = snap.getPropertyValue(propertyName);
+            Object propertyValueToSet = propertyValue;
+            if (propertyValue != null) {
+                try {
+                    Method setter = getSetterForProperty(clientsClass, propertyName);
+                    Class<?> targetType = setter.getParameterTypes()[0];
+                    if (targetType.isAssignableFrom(java.util.List.class)
+                            || targetType.isAssignableFrom(java.util.Set.class)) {
+                        Collection collection = (Collection) propertyValue;
+                        Set toBeDeleted = new HashSet<>();
+                        for (Object element : collection.toArray()){
+                            if (element instanceof InstanceId) {
+                                InstanceId instanceId = (InstanceId) element;
+
+                                Object targetObject = compileEntityForCommitId(
+                                        instanceId, snap.getCommitId());
+                                collection.add(targetObject);
+                                toBeDeleted.add(element);
+                            }
+                        }
+                        for (Object element : toBeDeleted) {
+                            collection.remove(element);
+                        }
+                    } else if (propertyValue instanceof InstanceId) {
+                        propertyValueToSet = compileEntityForCommitId(
+                                (InstanceId) propertyValue, snap.getCommitId());
+                    }
+                    setter.invoke(instance, propertyValueToSet);
+                } catch (Exception exc) {
+                    exc.printStackTrace();
+                }
+            }
+        }
+
+        return instance;
+    }
+
+    /**
+     * Programmatic way is chosen to find corresponding CdoSnapshot, however
+     * better practice would be to have the repository method for searching
+     *
+     * @param instanceId
+     * @param commitId
+     * @return
+     * @throws InstantiationException
+     * @throws IllegalAccessException
+     */
+    private Object compileEntityForCommitId(InstanceId instanceId, CommitId commitId) throws IllegalAccessException,
+            InstantiationException {
+
+        // TODO here the Javers function for loading the CdoSnapshot for the
+        // given commitId should be called, but the current Javers
+        // implementation doesn't provide it
+        // TODO assumes that limit = -1 means "all CdoSnapshots"
+        //List<CdoSnapshot> snapshots = javers.getStateHistory(convertInstanceIdToDTO(instanceId), -1); // load all
+        //List<CdoSnapshot> snapshots = javers.findSnapshots(QueryBuilder.byInstanceId(instanceId, instanceId.getgetBaseJavaClass() ).build());
+        List<CdoSnapshot> snapshots = javers.findSnapshots(QueryBuilder.byInstanceId(instanceId, instanceId.getClass()).build());
```

Comment:
fetching all snapshots could kill performance, we can think about adding queryByCommitId to JQL...